### DISCUSSION TITLE
Export eine PDF Datei pro Project

### DISCUSSION THREAD

**Discussion by jens-2305** [kimai/kimai#3195] [Questions & Answers]
Hallo,

ich würde im Export gerne die Möglichkeit schaffen das beim Export pro Projekt eine PDF Datei erstellt wird, diese anschließend auf dem Server gezipped werden und dann in die Response gegeben werden.
Mein Ansatzpunkt war hier die render Funktion im PDFRenderer.php leider gelingt es mir nicht die $query dort an zu passen.
Meine Annahme war das ich dieser $query die Abfrage finde und hier eine Iteration über die Projekte machen kann. 
Kann jemand helfen wie diese Iteration aussehen müsste ?

Vielen Dank

**Comment by kevinpapst** [comment]
Bau eine neue Klasse:
```php
final class ZipProjectRenderer implements RendererInterface
{
    public function getIcon(): string {
        return 'zip';
    }

    public function getTitle(): string {
        return 'ZIP';
    }

    public function render(array $exportItems, TimesheetQuery $query): Response {
        dd(__METHOD__);
    }

    public function getId(): string {
        return 'zip';
    }
}
```
Und dann in der Methode `public function render(array $exportItems, TimesheetQuery $query): Response;` einfach über die Objekte iterieren, in mehrere Listen (pro Projekt) aufteilen. Dann ZIP erstellen und das ZIP in die Response stecken.

Edit: und das Ergebnis gerne als Plugin teilen 👍 
Edit 2: Ganze Klasse ergänzt, damit der nächste nicht alle Kommentare lesen muss.

**Reply by jens-2305** [reply]
Ich habe eine PluginStruktur wie in der Doku angelegt.
Als zusätzliches Verzeichnis/Datei im Plugin habe ich `\Export\Base\ZipRenderer.php` angelegt.

Das Plugin sehe ich nun unter Erweiterungen, beim Export fehlt aber nun wieder der Button. Vermutlich liegt es an der `services.yaml` wie muss der Eintrag dort aussehen um den neuen Export zu ziehen und den Button anzuzeigen ?

**Reply by kevinpapst** [reply]
Ich glaube nicht, das es da was besonderes geben muss. 
Ich kann jetzt auch nur raten ehrlich gesagt:

```yaml
services:
    _defaults:
        autowire: true
        autoconfigure: true
        public: false

    KimaiPlugin\ZipExportBundle\:
        resource: '../../*'
        exclude: '../../{Resources}'
```

Aber mehr kann ich nur sagen, wenn Du es als Repo teilst, damit ich reinschauen kann.
Lädst Du die services.yaml denn auch?

**Comment by jens-2305** [comment]
Danke stand auf dem Schlauch. Konnte die Lösung umsetzen. 
Ein Plugin habe ich noch nie gemacht, ich schaue mal ob ich das hinbekomme und melde mich dann gerne an dieser Stelle.
Danke für deine Hilfe.

**Reply by kevinpapst** [reply]
Ich wette viele würden sich darüber freuen, wenn Du den Code teilst. Wenn nicht als Plugin, dann vielleicht einfach hier als Anleitung?

**Reply by jens-2305** [reply]
Na sagen wir mal wenn dann richtig ;-)

Sollte ich das nicht hinbekommen dann teile ich hier den Code, versprochen.

**Reply by jens-2305** [reply]
Wenn ich im 

src/Export/Base/PDFRenderer.php

die Funktion 
render(array $exportItems, TimesheetQuery $query): Response 

"editiere" bekomme ich die Lösung hin. 

Allerdings gelingt es mir nicht wenn ich einen neuen anlege unter src/Export/Base/ZipProjectRenderer.php
Es erscheint kein Button auf der Exportseite, bin ich an der falschen Stelle oder habe ich was übersehen ?

In der Doku steht das der Button automatisch angezeigt werden soll wenn man den Cache refreshed. Irgendwas
scheint aber zu fehlen.

**Reply by kevinpapst** [reply]
Funktioniert wie es in der Doku steht: https://www.kimai.org/documentation/developers.html#adding-export-renderer
```
namespace App\Export\Base;

use App\Export\RendererInterface;
use Symfony\Component\HttpFoundation\Response;

final class ZipProjectRenderer implements RendererInterface
{
    public function getIcon(): string {
        return 'zip';
    }

    public function getTitle(): string {
        return 'ZIP';
    }

    public function render(array $exportItems, TimesheetQuery $query): Response {
        dd(__METHOD__);
    }

    public function getId(): string {
        return 'zip';
    }
}
```
<img width="336" alt="Bildschirmfoto 2022-03-17 um 16 53 59" src="https://user-images.githubusercontent.com/533162/158840256-c4c45839-67bd-4354-a6f8-406f3fae96ee.png">

**Reply by jens-2305** [reply]
Danke, hatte wie in deiner ersten Antwort `implements ExportRendererInterface` verwendet, damit gings nicht, wenn ich nun `implements RendererInterface` verwendet funktioniert es, danke für die Hilfe ich bleibe dran ;-)

**Reply by kevinpapst** [reply]
Ah, ich hab die Antwort angepasst, damit der nächste nicht auch drüber stolpert.
Tja, wenn man nicht in seine eigene Doku schaut 😁
